File: leet-code/200. Number of Islands/200_number_of_islands.py
def num_islands(grid):
    def dfs_recursion(y, x):
        visited[y][x] = True

        for k in range(4):
            y_idx = y + y_value[k]
            x_idx = x + x_value[k]

            if 0 > y_idx or y_idx >= m or 0 > x_idx or x_idx >= n:
                continue

            if visited[y_idx][x_idx]:
                continue

            if grid[y_idx][x_idx] == '0':
                continue

            dfs_recursion(y_idx, x_idx)

    m = len(grid)
    n = len(grid[0])
    y_value = [-1, 0, 1, 0]
    x_value = [0, 1, 0, -1]
    visited = [[False] * n for _ in range(m)]
    cnt = 0

    for i, row in enumerate(grid):
        for j, r in enumerate(row):
            if visited[i][j]:
                continue

            if grid[i][j] == '0':
                continue

            cnt += 1
            dfs_recursion(i, j)

    return cnt


print(num_islands([
    ["1", "1", "1", "1", "0"],
    ["1", "1", "0", "1", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "0", "0", "0"]
]))

print(num_islands([
    ["1", "1", "0", "0", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "1", "0", "0"],
    ["0", "0", "0", "1", "1"]
]))

File: leet-code/200. Number of Islands/200_number_of_islands.c
static void dfs_recursion(char** grid, const int gridSize, int* gridColSize, const int i, const int j)
{
    if (0 > i || i >= gridSize || 0 > j || j >= gridColSize[i] || grid[i][j] != '1') {
        return;
    }
    grid[i][j] = '0';
    dfs_recursion(grid, gridSize, gridColSize, i - 1, j);
    dfs_recursion(grid, gridSize, gridColSize, i, j + 1);
    dfs_recursion(grid, gridSize, gridColSize, i + 1, j);
    dfs_recursion(grid, gridSize, gridColSize, i, j - 1);
}

int numIslands(char** grid, int gridSize, int* gridColSize)
{
    int cnt = 0;
    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridColSize[i]; j++) {
            if (grid[i][j] == '1') {
                cnt += 1;
                dfs_recursion(grid, gridSize, gridColSize, i, j);
            }
        }
    }
    return cnt;
}
